### PR TITLE
use common sql interfaces

### PR DESCRIPTION
**Pull request by prolic**
See also here: https://github.com/prolic/sql
(We would need to move the repo to amphp organization)

If I missed something, let me know please.

### CODE REVIEW

**Comment by prolic** [comment]
see also https://github.com/amphp/postgres/pull/14

**Comment by Sevavietl** [comment]
Only a couple of days ago I was thinking about switching between MySQL and Postgres, and how problematic that would be without the common interface.

File: src/CommandResult.php [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 namespace Amp\Mysql;
 
-final class CommandResult {
```
Comment (kelunik) [inline src/CommandResult.php]: I prefer aliased imports instead, e.g. `use Amp\Sql` → `implements Sql\CommandResult` or `use Amp\Sql\CommandResult as SqlCommandResult`.

File: src/FailureException.php [hunk 2]
```diff
@@ -2,5 +2,5 @@
 
 namespace Amp\Mysql;
 
-class FailureException extends \Exception {
+class FailureException extends \Amp\Sql\FailureException {
 }
```
Comment (kelunik) [inline src/FailureException.php]: Do we need the extension here and can't use the base class directly?

File: src/Transaction.php [hunk 3]
```diff
@@ -3,14 +3,10 @@
 namespace Amp\Mysql;
 
 use Amp\Promise;
+use Amp\Sql\Executor;
 use function Amp\call;
 
-final class Transaction implements Executor, Operation {
```
Comment (kelunik) [inline src/Transaction.php]: Shouldn't `Sql\Transaction` extend `Executor` instead?

File: src/FailureException.php (hunk 2 above)
Comment (bwoebi) [reply src/FailureException.php]: Yes, we do. To just catch any SQL errors easily.

File: src/ResultSet.php [hunk 4]
```diff
@@ -7,9 +7,10 @@
 use Amp\Iterator;
 use Amp\Producer;
 use Amp\Promise;
+use Amp\Sql\Operation;
 use Amp\Success;
 
-final class ResultSet implements Iterator, Operation {
+final class ResultSet implements Operation, \Amp\Sql\ResultSet {
```
Comment (bwoebi) [inline src/ResultSet.php]: \Amp\Sql\ResultSet should directly extend Operation instead of having to implement both interfaces here

**Comment by prolic** [comment]
maybe we can also unify the `Collector` interface?

**Comment by prolic** [comment]
@kelunik changes done, I also invited to as collaborator, so you can transfer the repository.

**Comment by bwoebi** [comment]
Why did you rename the Connector.php file to Collector?

Also, unifying Connectors sounds like a good idea, the basic connection configuration is identical. However it probably should be of type `string | ConnectionConfig` then (with ConnectionConfig being a non-final class extended in each of the repositories with mysql/pg specific configuration).

**Comment by prolic** [comment]
@bwoebi ups, my bad, fixed

File: src/FailureException.php (hunk 2 above)
Comment (kelunik) [reply src/FailureException.php]: For any SQL error it's fine to use `Amp\Sql\FailureException`, no?

**Comment by kelunik** [comment]
@prolic Seems like I only have read / write access, but no admin access, which is required to transfer a repository.

**Comment by prolic** [comment]
@kelunik I transferred the repo to your private account, then you can move to amphp. Admin rights are not possible for private account's repos (I think).

**Comment by prolic** [comment]
@bwoebi about the connection options as argument for connector: sounds good, I'll try to find some time tomorrow to add this change.

**Comment by prolic** [comment]
@kelunik @trowski I also refactored the connection object, please review and let me know if I missed something and see also postgres-PR.

**Comment by trowski** [comment]
Looks good, just a couple of things:

`Connection::connect()` should not be part of the interface, as postgres connection objects do not have this method. As for switching away from the static constructor for `Connection` in this library, I don't feel strongly either way, but the current code will require some changes in `Processor` because the connection is in an unusable state until `connect()` is called. I probably went with the static constructor because it just seemed simpler. The static constructor also had a nice duality with the postgres library.

You repeat methods in interfaces that inherit, such as `Connection` and `Pool`. I feel this is confusing as I lose track of where the method is originally defined.

File: test/PoolTest.php [hunk 5]
```diff
@@ -31,8 +31,7 @@ protected function createPool(array $connections): Pool {
         $connector = $this->createMock(Connector::class);
         $connector->method('connect')
             ->will($this->returnCallback(function () use ($connections): Promise {
-                static $count = 0;
-                return new Success($connections[$count++ % \count($connections)]);
+                return new Success();
```
[the remark this answers is not in the document]
Comment (prolic) [reply test/PoolTest.php]: the promise on connect only resolves, there is no return value, as the connect method is no more static.

**Comment by prolic** [comment]
@trowski see f.e. here: https://github.com/prolic/postgres/blob/c503e0bd516b923d9e61bc23915e3bffec50859b/src/PgSqlConnection.php#L23

**Comment by prolic** [comment]
@trowski `You repeat methods in interfaces that inherit, such as Connection and Pool.` - Where exactly? I will fix it.

File: src/CommandResult.php (hunk 1 above)
Comment (prolic) [reply src/CommandResult.php]: done

**Comment by prolic** [comment]
Another note, before we merge mysql / postgres changes, we should move sql repo into amphp organization and create an entry in packagist. Then we can re-run the travis tests for mysql / postgres and see if something is broken somewhere.

**Comment by prolic** [comment]
mmh.. thinking again about the connect method. Maybe I can make put `final public static function connect(ConnectionConfig $config, CancellationToken $token = null): Promise` in the interface as static constructor. I can build then the rest internally. For Mysql this means that `$socket = yield Socket\connect($config->connectionString(), $connectContext, $token);` is called within this static constructor, no need to pass a socket connection. Thoughts?

**Comment by trowski** [comment]
@prolic `transaction()` is repeated in [`Pool`](https://github.com/kelunik/sql/blob/master/src/Pool.php#L39).
`isAlive()`, `close()`, `query()`, `transaction()`, `prepare()`, and `execute()` are repeated in [`Connection`](https://github.com/kelunik/sql/blob/master/src/Connection.php).

> before we merge mysql / postgres changes, we should move sql repo into amphp organization and create an entry in packagist.

Of course. Just need @kelunik to transfer it.

> Maybe I can make put `final public static function connect(ConnectionConfig $config, CancellationToken $token = null): Promise` in the interface as static constructor.

Yeah, that should be fine. We can mock the socket when creating the `Processor` object passed to `Connection` for testing purposes, so the socket can be created in `connect()` as you suggest.
Then just revert the changes you made with the `connect()` method in postgres as well (except using a `ConnectionConfig` object in postgres instead of a string of course).

**Comment by kelunik** [comment]
I've transferred the repository and added it to Packagist.

**Comment by kelunik** [comment]
Travis CI and Coveralls are enabled as well now.

**Comment by prolic** [comment]
Do I still have write access?

**Comment by kelunik** [comment]
You should already have an invitation for write access, yes. :-)

**Comment by prolic** [comment]
Nice. Will finish the rest tomorrow.

File: .php_cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+$config = new Amp\CS\Config\Amp();
```
Comment (kelunik) [inline .php_cs]: It doesn't really matter, but I'd prefer `Amp\CodeStyle\Config` here.

**Comment by prolic** [comment]
rebased to resolve conflicts

**Comment by prolic** [comment]
@kelunik travis didn't fetch latest amphp/sql changes, it's testing old commits, thus travis failed. any idea why?

**Comment by prolic** [comment]
next build again:
`- Installing amphp/sql (dev-master 57d7276): Loading from cache`

but latest commit is `9d31f9be697421f46cc97489e9f2e7862c333b61`

**Comment by prolic** [comment]
@kelunik maybe amphp/sql has not the packagist auto update enabled?

**Comment by kelunik** [comment]
@prolic The Packagist service is enabled, but sure why it's not auto-updated. Manually updated it on Packagist now.

**Comment by prolic** [comment]
Ok, can you restart mysql travis build? So I can see if I missed something.

**Comment by prolic** [comment]
https://packagist.org/packages/amphp/sql is not updated, thus 7.0 tests fail on travis.
7.1 + 7.2 tests have segmentation fault on travis.

**Comment by trowski** [comment]
@prolic Updated Pool in amphp/sql. Feel free to update this PR based on that. Something came up that needs my attention today, but I should have time to look at this again tonight.

**Comment by prolic** [comment]
@trowski @kelunik updated

**Comment by prolic** [comment]
@trowski In AbstractPool class we have this: https://github.com/amphp/sql/blob/master/src/AbstractPool.php#L75

But there is no `doPrepare` method anymore. Hence all tests fail.

**Comment by trowski** [comment]
Oh yeah… *thats* why `prepare()` was separated like that. Fixed.

**Comment by prolic** [comment]
Can you restart Travis build please?

**Comment by bwoebi** [comment]
Nothing against purism, but `$db = yield Mysql\connect(Mysql\ConnectionConfig::parseConnectionString("host=".DB_HOST.";user=".DB_USER.";pass=".DB_PASS.";db=".DB_NAME));` is just ugly.

**Comment by prolic** [comment]
We can import that directly if you prefer

**Comment by bwoebi** [comment]
Yeah, mysql\connect should try to parse parameter as connection string if's not a config

**Comment by prolic** [comment]
I don't like this type juggling stuff. If it's an object do this, if it's a
string do that, etc. Better to add another function like
Mysql\connectFromString

**Comment by prolic** [comment]
Yes

**Comment by prolic** [comment]
@trowski what needs to be done to get this merged?

**Comment by trowski** [comment]
@prolic I don't have time to look in detail right now, but it needs to be updated to the latest changes in amphp/sql if it isn't already. Use amphp/postgres for guidance if there's implementation questions or ask in IRC. 😄

**Comment by prolic** [comment]
Ok, I'll check this weekend
Sascha-Oliver Prolic


Am Do., 13. Sep. 2018 um 22:48 Uhr schrieb Aaron Piotrowski <
notifications@github.com>:

File: src/ConnectionConfig.php [hunk 7]
```diff
@@ -201,20 +153,20 @@ public function getCollation(): string
         return $this->collate;
     }
 
-    public function withCharset(string $charset, string $collate)
+    public function withCharset(string $charset = self::DEFAULT_CHARSET, string $collate = self::DEFAULT_COLLATE): self
```
Comment (kelunik) [inline src/ConnectionConfig.php]: Not sure whether default parameters make sense here.

Comment (prolic) [reply src/ConnectionConfig.php]: At least the second is good, so I can set charset without passing new collate. But we can also remove it, it you prefer.

**Comment by prolic** [comment]
Some tests are still failing. Unfortunately I'm out of time for today. Feel free to pick up my work and continue on it.

File: src/PooledResultSet.php [hunk 8]
```diff
@@ -18,4 +19,9 @@ public function __construct(ResultSet $result, callable $release)
         parent::__construct($result, $release);
         $this->result = $result;
     }
+
+    public function nextResultSet(): Promise
```
Comment (kelunik) [inline src/PooledResultSet.php]: Should we rename this to `advanceResultSet()`?

Comment (prolic) [reply src/PooledResultSet.php]: Why is this method even needed? We don't have something like it in amp\postgres. Maybe should should hide this somehow as implementation detail that you don't care about and is not part of public api.

**Comment by prolic** [comment]
and segmentation fault on travis!

**Pull request merged**